src/marlin_stubs/sdcard/M20-M30_M32-M34.cpp

```diff
@@ -1,22 +1,80 @@
 #include <dirent.h>
 
 #include "../../lib/Marlin/Marlin/src/gcode/gcode.h"
+#include "libs/hex_print_routines.h"
 #include "marlin_server.hpp"
 #include "media.h"
 #include "marlin_vars.hpp"
 
-// M20 - List SD card
-void GcodeSuite::M20() {
-    SERIAL_ECHOLNPGM(MSG_BEGIN_FILE_LIST);
+struct TimeFlags {
+    bool print_lfn : 1;
+    bool print_timestamps : 1;
+};
+
+static void time_to_m20(const struct stat stat_buf, uint16_t *m20_date, uint16_t *m20_time) {
+    time_t t = stat_buf.st_mtim.tv_sec;
+    struct tm lt;
+    localtime_r(&t, &lt);
+    *m20_date = (lt.tm_year + 1900 - 1980) << 9 | (lt.tm_mon + 1) << 5 | lt.tm_mday;
+    *m20_time = lt.tm_hour << 11 | lt.tm_min << 5;
+    *m20_time |= int((lt.tm_sec - (lt.tm_sec % 2)) / 2);
+}
+
+static void print_file_info(const char *const path, const char *const long_path, const struct TimeFlags tf) {
+    // &path[5] and &long_path[5], because we skip the /usb/ prefix on print as it is only used internally
+    SERIAL_ECHO(&path[5]);
+    SERIAL_CHAR(' ');
+    struct stat stat_buf;
+    int rc = stat(path, &stat_buf);
+    if (rc == 0) {
+        SERIAL_ECHO(stat_buf.st_size);
+        if (tf.print_timestamps) {
+            uint16_t m20_date;
+            uint16_t m20_time;
+            time_to_m20(stat_buf, &m20_date, &m20_time);
+            SERIAL_ECHOPGM(" 0x");
+            print_hex_word(m20_date);
+            print_hex_word(m20_time);
+        }
+        if (tf.print_lfn) {
+            SERIAL_CHAR(' ');
+            SERIAL_CHAR('"');
+            SERIAL_ECHO(&long_path[5]);
+            SERIAL_CHAR('"');
+        }
+    }
+    SERIAL_EOL();
+}
+
+static void print_listing(const char *const prefix, const char *const prefix_long, const struct TimeFlags tf) {
     DIR *dir;
-    dir = opendir("/usb/");
+    dir = opendir(prefix);
     if (dir != NULL) {
         struct dirent *entry;
         while ((entry = readdir(dir)) != NULL && entry->d_name[0]) {
-            SERIAL_ECHOLN(entry->d_name);
+            char path[5 + sizeof(entry->d_name) * 2];
+            snprintf(path, sizeof(path), "%s/%s", prefix, entry->d_name);
+            char long_path[5 + sizeof(entry->lfn) * 2];
```

Comment:
I still don't understand these buffer sizes. Why `*2` ? We are only supporting root directory, right? I would imagine we need "/usb/" + respective filename (sfn or lfn) + possibly a null terminator. What am I missing?

Reply:
The idea was that sizeof(entry->d_name) has the same size as prefix. But yeah, I've changed that completely because it wasn't clean and safe. Should be fine now.